server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -316,17 +322,18 @@ private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOExc
     }
 
     private boolean upload(long primaryTerm, long generation, long maxSeqNo) throws IOException {
-        // During primary relocation (primary-primary peer recovery), both the old and the new primary have engine
-        // created with the RemoteFsTranslog. Both primaries are equipped to upload the translogs. The primary mode check
-        // below ensures that the real primary only is uploading. Before the primary mode is set as true for the new
-        // primary, the engine is reset to InternalEngine which also initialises the RemoteFsTranslog which in turns
-        // downloads all the translogs from remote store and does a flush before the relocation finishes.
-        if (primaryModeSupplier.getAsBoolean() == false) {
-            logger.debug("skipped uploading translog for {} {}", primaryTerm, generation);
+        // During primary relocation, both the old and new primary have engine created with RemoteFsTranslog and having
+        // ReplicationTracker.primaryMode() as true. However, before we perform the `internal:index/shard/replication/segments_sync`
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using startedPrimarySupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (startedPrimarySupplier.getAsBoolean() == false || syncPermit.tryAcquire(SYNC_PERMIT) == false) {
+            logger.debug("skipped uploading translog for {} {} syncPermits={}", primaryTerm, generation, syncPermit.availablePermits());
             // NO-OP
             return true;
         }
```

Comment:
should we move this to `prepareAndUpload` ? There is no point creating another generation locally , if we can't rotate it .

Reply:
Do you mean to say `upload` instead of `roatate`?

In context of upload, I will check this a bit more if we can move this check to prepareAndUpload and I think we might have discussed this in the older PR. Also, we might need similar change for segments i.e. prevent segments or segment infos creation if we can't upload.

Reply:
This would be tricky for segments as the creation of local segments in not in our control as of now , unlike translog .